roles/postgres/files/rotate.py

```diff
@@ -0,0 +1,109 @@
+#!/usr/bin/python
+
+"""Rotate backup files in a given directory.
+
+Implements a primitive FIFO backup rotation strategy by keeping N most recent backup files. The
+order is defined by sorting the file names lexicographically: the files that appear later in
+the sorted list are considered to be newer.
+"""
+
+import argparse
+import dataclasses
+import pathlib
+
+
+@dataclasses.dataclass(order=True)
+class Backup:
+    path: pathlib.Path
+    size: int
+
+
+@dataclasses.dataclass
+class Args:
+    keep: int
+    dir: pathlib.Path
+    pattern: str
+    nodry_run: bool
+
+
+def non_negative_int(str_value: str) -> int:
+    value = int(str_value)
+    if value < 0:
+        raise argparse.ArgumentTypeError(f"Value must be non-negative: {value} < 0")
+
+    return value
+
+
+def parse_args() -> Args:
+    parser = argparse.ArgumentParser(description=__doc__)
+    parser.add_argument(
+        "-n",
+        "--keep",
+        required=True,
+        type=non_negative_int,
+        help="Keep this many most recent backup files",
+    )
+    parser.add_argument(
+        "-d", "--dir", type=pathlib.Path, help="Path to the directory with backup files"
+    )
+    parser.add_argument(
+        "-p",
+        "--pattern",
+        type=str,
+        default="*",
+        help="Only consider files that match this glob pattern",
+    )
+    parser.add_argument(
+        "--nodry_run",
+        action="store_true",
+        help="Actually remove the rotated files",
+    )
+
+    return Args(**vars(parser.parse_args()))
```

Comment:
Don't laugh, I was wondering, what can we do to preserve the original `Namespace` object yet preserve typing. :smile: 

```python
import argparse
import typing

class Args(typing.Protocol):
    foo: str
    bar: str

def get_args() -> Args:
    parser = argparse.ArgumentParser()
    parser.add_argument("foo")
    parser.add_argument("--bar")
    namespace = parser.parse_args()
    return typing.cast(Args, namespace)


args = get_args()
args.bar
```

It's not like I'm suggesting this necessarily, but just wanted to share it with you.

Reply:
Actually, I think it's not bad! That's essentially what I was trying to emulate with a dataclass. I didn't know about `typing.cast()`! Thank you for the suggestion! :pray: